scipy/sparse/csgraph/_flow.pyx

```diff
@@ -407,3 +426,207 @@ cdef ITYPE_t[:] _edmonds_karp(
+                                        heads, flows, rev_edge_ptr,
+                                        min(flow, capacities[e]))
+            if result_flow:
+                capacities[e] -= result_flow
+                capacities[rev_edge_ptr[e]] += result_flow
+                flows[e] += result_flow
+                flows[rev_edge_ptr[e]] -= result_flow
+                return result_flow
+        progress[current] += 1
+    return 0
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef ITYPE_t[:] _dinic(
+        ITYPE_t[:] edge_ptr,
+        ITYPE_t[:] tails,
+        ITYPE_t[:] heads,
+        ITYPE_t[:] capacities,
+        ITYPE_t[:] rev_edge_ptr,
+        ITYPE_t source,
+        ITYPE_t sink):
+    """Solves the maximum flow problem using the Dinic's algorithm.
+
+    This assumes that for every edge in the graph, the edge in the opposite
+    direction is also in the graph (possibly with capacity 0).
+
+    Parameters
+    ----------
+    edge_ptr : memoryview of length :math:`|V| + 1`
+        For a given vertex v, the edges whose tail is ``v`` are those between
+        ``edge_ptr[v]`` and ``edge_ptr[v + 1] - 1``.
+    tails : memoryview of length :math:`|E|`
+        For a given edge ``e``, ``tails[e]`` is the tail vertex of ``e``.
+    heads : memoryview of length :math:`|E|`
+        For a given edge ``e``, ``heads[e]`` is the head vertex of ``e``.
+    capacities : memoryview of length :math:`|E|`
+        For a given edge ``e``, ``capacities[e]`` is the capacity of ``e``.
+    rev_edge_ptr : memoryview of length :math:`|E|`
+        For a given edge ``e``, ``rev_edge_ptr[e]`` is the edge obtained by
+        reversing ``e``. In particular, ``rev_edge_ptr[rev_edge_ptr[e]] == e``.
+    source : int
+        The source vertex.
+    sink : int
+        The sink vertex.
+
+    Returns
+    -------
+    flows : memoryview of length :math:`|E|`
+        The residual graph with respect to a maximum flow.
+
+    """
+    cdef ITYPE_t n_verts = edge_ptr.shape[0] - 1
+    cdef ITYPE_t n_edges = capacities.shape[0]
+    cdef ITYPE_t ITYPE_MAX = np.iinfo(ITYPE).max
+
+    cdef ITYPE_t[:] levels, progress
+    cdef ITYPE_t[:] q = np.empty(n_verts, dtype=ITYPE)
+    cdef ITYPE_t[:] flows = np.zeros(n_edges, dtype=ITYPE)
+    cdef ITYPE_t flow
+
+    cdef ITYPE_t max_flow = 0
+    while True:
+        levels = np.full(n_verts, -1, dtype=ITYPE)
+        if not _build_level_graph(edge_ptr, source, sink,
+                                 capacities, heads, levels, q):
+            break
+        progress = np.full(n_verts, -1, dtype=ITYPE)
+        while True:
+            flow = _augment_paths(edge_ptr, source, sink, 
+                                 progress, levels, capacities,
+                                 heads, flows, rev_edge_ptr, ITYPE_MAX)
+            if flow:
+                max_flow += flow
+            else:
+                break
+    return flows
```

Comment:
I guess it's possible to make sure interaction with the GIL are minimized there.

It might be possible to allocate `levels`, `progress`, `q`,`flows` once initially with numpy python API (which necessitate the GIL) and then make sure the rest of the code runs without it.

Also:
```python
        levels = np.full(n_verts, -1, dtype=ITYPE)
```
and
```python
        progress = np.full(n_verts, -1, dtype=ITYPE)
```

reallocate buffers everytimes, and it might be faster flushing their content manually.

Reply:
> it might be faster flushing their content manually.

Thanks for pointing out. These two calls have been my concern too. Should we use manual loops to reset all the values to -1 or by using `memset`?

Reply:
> Should we use manual loops to reset all the values to -1 or by using memset?

[memset should only be used on bytes or strings](https://linux.die.net/man/3/memset) otherwise, the behavior is undefined.

It's possible to use it to set buffers' elements to 0, but it's a working edge case as numerical zeros are represented using null bytes: using it with another constant (e.g -1 here) would give wrong results, at least [that's the case for floats](https://mackiemathew.wordpress.com/2015/06/30/dont-use-memset-for-initializing-floats-or-doubles/).

For loops aren't elegant but should be mapped effectively to efficient C code.